File: test/test_hector.cpp
#include "Hector/Core/Exception.h"
#include "Hector/Core/ParticleStoppedException.h"

#include "Hector/IO/TwissHandler.h"
#include "Hector/Beamline.h"

#include "Hector/Propagator/Propagator.h"
#include "Hector/Utils/BeamProducer.h"
#include "Hector/Utils/ArgsParser.h"

#include <CLHEP/Random/RandGauss.h>

using namespace std;

int main(int argc, char* argv[]) {
  string twiss_file, ip;
  double min_s, max_s;
  unsigned int num_part = 100;
  bool shoot;
  hector::ArgsParser(argc,
                     argv,
                     {{"twiss-file", "beamline Twiss file", &twiss_file, 'i'}},
                     {
                         {"interaction-point", "name of the interaction point", "IP5", &ip, 'c'},
                         {"min-s", "minimum arc length s to parse (m)", 0., &min_s},
                         {"max-s", "maximum arc length s to parse (m)", 250., &max_s},
                         {"num-part", "number of particles to shoot", 10, &num_part, 'n'},
                         {"simulate", "simulate a beam propagation", false, &shoot, 's'},
                     });

  hector::io::Twiss parser(twiss_file.c_str(), ip.c_str(), max_s, min_s);
  parser.printInfo();
  cout << "+---------------------+--------------------+----------------------" << endl;
  cout << hector::Form("| %-19s | %-18s | %20s|", "Name", "Type", "Position along s (m)") << endl;
  cout << "+---------------------+--------------------+----------------------" << endl;
  for (const auto& elem : parser.beamline()->elements()) {
    //if ( elem->type() == hector::element::aDrift ) continue;
    cout << hector::Form("|%20s | %-18s [ ",
                         elem->name().c_str(),
                         (elem->type() != hector::element::aDrift) ? elem->typeName().c_str() : "");
    string pos_ini = hector::Form("%7s", hector::Form("%#0.3f", elem->s()).c_str());
    if (elem->length() > 0.)
      cout << hector::Form("%.7s → %7s m", pos_ini.c_str(), hector::Form("%#0.3f", elem->s() + elem->length()).c_str());
    else
      cout << hector::Form("%17s m", pos_ini.c_str());
    cout << " ]" << endl;
  }
  cout << "+---------------------+-------------------------------------------" << endl;
  //parser.beamline()->dump();

  cout << "beamline matrix at s = " << max_s
       << " m: " << parser.beamline()->matrix(0., hector::Parameters::get()->beamParticlesMass(), +1) << endl;

  if (shoot) {
    hector::Propagator prop(parser.beamline());
    //parser.beamline()->dump();

    hector::beam::GaussianParticleGun gun;
    gun.smearEnergy(hector::Parameters::get()->beamEnergy(), hector::Parameters::get()->beamEnergy() * 0.);
    //hector::beam::TXscanner gun( num_part, hector::Parameters::get()->beamEnergy(), 0., 1. );
    map<string, unsigned int> stopping_elements;
    for (unsigned int i = 0; i < num_part; ++i) {
      hector::Particle p = gun.shoot();
      p.setCharge(+1);
      try {
        prop.propagate(p, 203.826);
      } catch (hector::ParticleStoppedException& e) {
        stopping_elements[e.stoppingElement()->name()]++;
        continue;
      } catch (hector::Exception& e) {
        e.dump();
      }
    }

    ostringstream os;
    os << "Summary\n\t-------";
    for (const auto& el : stopping_elements)
      os << hector::Form("\n\t*) %.2f%% of particles stopped in %s", 100. * el.second / num_part, el.first.c_str());
    PrintInfo << os.str() << ".";
  }

  return 0;
}

File: test/test_hector.py
import pyhector as hector
from ROOT import gROOT, TCanvas, TH2D, TH1D
#from pympler.tracker import SummaryTracker

gROOT.SetBatch(1)

#tracker = SummaryTracker()

### beamline retrieval part

parser = hector.Twissparser('data/twiss/twiss_2016-prels2_ir5b1_6p5tev.tfs', 'IP5', 1)
#print [(e.name, e.type) for e in parser.beamline.elements]
#for elem in parser.beamline().elements():
#    print elem.s, elem.name
coll = parser.beamline.getElement('TANC.4R5')
print coll, coll.aperture.position, coll.matrix(0.)

#exit(0)
prop = hector.Propagator(parser.beamline)

params = hector.Parameters.get()

### particles generation

hm = TH2D('hitmap', ';x (mm);y (mm)', 200, -10., 10., 200, -10., 10.)
hm0 = TH2D('hitmap_s0', ';T_{x} (mrad);T_{y} (mrad)', 200, -1., 1., 200, -1., 1.)
h_numelems = TH1D('num_elems', ';Number of elements before particle is stopped', 100, 0., 100. )

sv = hector.StateVector()
p_energy = params.beamEnergy*0.9
#gun = hector.GaussianParticleGun({'Emin': 6400., 'Emax': 6500., 'Txmax': 10.e-6, 'Tymax': 10.e-6})
gun = hector.GaussianParticleGun({'Emin':p_energy, 'Emax':p_energy, 'q': 1, 'm': params.beamParticlesMass})
gun.smearX(0., 10.e-6)
gun.smearY(0., 10.e-6)
gun.smearTx(191.541e-6, 20.e-6)
gun.smearTy(0., 20.e-6)

for i in range(50000):
    try:
        part = gun.shoot()
        pos0 = part.positions[0]
        hm0.Fill(pos0.Tx*1.e3, pos0.Ty*1.e3)
        try:
            prop.propagate(part, 250.)
            sv = part.stateVectorAt(210.)
            hm.Fill(sv.x*1.e3, sv.y*1.e3)
        except hector.ParticleStoppedException, e:
            stop_elem = e.stoppingElement
            print 'stopped in', stop_elem.name, stop_elem.type, 'at s =', stop_elem.s, '-', stop_elem.s+stop_elem.length, 'm after', len(part.positions), 'element(s)'
            print 'last state vector at s =', part.positions.keys()[-1], 'm:', part.positions[part.positions.keys()[-1]]
            h_numelems.Fill(len(part.positions))
            #break
            continue
    except KeyboardInterrupt:
        break

#tracker.print_diff()

#mat = parser.beamline.matrix(10.)
#print mat

c = TCanvas()
c.Divide(2,2)
c.cd(1)
hm.Draw('colz')
c.cd(2)
h_numelems.Draw()
c.SaveAs('test.png')

